spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/SparkWriteConf.java

```diff
@@ -174,12 +174,17 @@ public long targetDataFileSize() {
         .parse();
   }
 
-  public boolean fanoutWriterEnabled() {
+  public boolean useFanoutWriter(SparkWriteRequirements writeRequirements) {
+    boolean defaultValue = !writeRequirements.hasOrdering();
```

Comment:
Should this be slightly more complex? If the distribution mode was explicitly set to `none`, for example, then this is not considered safe. I think we should ensure that there is no ordering and that there is a requested distribution.

Reply:
I reasoned a bit about this [here](https://github.com/apache/iceberg/pull/8621#issuecomment-1732211168). I am open to that.

Reply:
Actually, we may want to keep it this way.

Another use case that may benefit from the current approach is SPJ. There, users set the distribution mode to none. They would get a super expensive local sort and spill without explicitly enabling fanout writers. I think that's a more realistic use case than setting to `none` and generating tons of files per task. If the user sets it to `none` explicitly, they are probably OK with the number of produced files, which hints it is not a crazy number. So why do a local sort for them?

We generally preferred safe options vs more performant and it meant more configs required to improve the performance. I'd say let's turn that around.

Reply:
I think the argument against it is that this would blow up memory. A local sort is at least memory-safe but opening hundreds of Parquet files at once is not and could cause an OutOfMemoryError.

Reply:
Okay, Anton and I spoke directly and concluded that this is probably a good idea after all. Distributing with `none` is never done by default and must be explicitly set. In that case, I think it's reasonable to need to tune a bit more and possibly add a local sort to handle this case.